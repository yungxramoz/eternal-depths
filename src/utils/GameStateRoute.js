import React from 'react'
import { useSelector } from 'react-redux'
import { Navigate, Outlet, useLocation } from 'react-router-dom'
import GAME_STATE from '../constants/game-state'
import ROUTE from '../constants/routes'

const GameStateRoute = () => {
  const gameState = useSelector((state) => state.game.gameState)
  const location = useLocation()

  const idlePaths = [ROUTE.HOME, ROUTE.NEW_CHARACTER, ROUTE.LEADERBOARD]
  const overPaths = [ROUTE.GAME_OVER]
  const wonPaths = [ROUTE.GAME_WON]
  const playingPaths = [ROUTE.DUNGEON]

  const navigateToPath = (path) => {
    return <Navigate to={path} />
  }

  if (gameState === GAME_STATE.IDLE && !idlePaths.includes(location.pathname)) {
    return navigateToPath(ROUTE.HOME)
  }
  if (gameState === GAME_STATE.OVER && !overPaths.includes(location.pathname)) {
    return navigateToPath(ROUTE.GAME_OVER)
  }
  if (gameState === GAME_STATE.WON && !wonPaths.includes(location.pathname)) {
    return navigateToPath(ROUTE.GAME_WON)
  }
  if (
    gameState === GAME_STATE.PLAYING &&
    !playingPaths.includes(location.pathname)
  ) {
    return navigateToPath(ROUTE.DUNGEON)
  }

  return <Outlet />
}

export default GameStateRoute
