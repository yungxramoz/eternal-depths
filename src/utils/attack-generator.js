import { ATTACK } from '../constants/attack-type'

const getRandomAttack = () => {
  const attacks = Object.values(ATTACK)
  return attacks[Math.floor(Math.random() * attacks.length)]
}

export const generateAttack = (attack) => {
  if (!attack) {
    attack = getRandomAttack()
  }
  return attack
}
