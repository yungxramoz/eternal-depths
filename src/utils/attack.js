import ATTACK_RESULT from '../constants/attack-result'
import {
  criticalChance,
  damagePoints,
  evasionChance,
  toHitChance,
} from './stats'

export const calculateDamage = (
  attack,
  sourceStats,
  targetStats,
  isEncounter = false,
) => {
  let damage = determineDamage(sourceStats, isEncounter)
  damage = Math.max(damage + attack.additionalDamage, 1) // additional damage can be negative
  if (!attack.safeHit) {
    const isHit = determineHit(sourceStats, targetStats)
    if (!isHit) {
      return {
        result: ATTACK_RESULT.MISSED,
        damage: 0,
      }
    }

    const isEvaded = determineEvade(targetStats)
    if (isEvaded) {
      return {
        result: ATTACK_RESULT.EVADED,
        damage: 0,
      }
    }

    const isCritical = determineCritical(sourceStats)
    if (isCritical) {
      return {
        result: ATTACK_RESULT.CRITICAL,
        damage: Math.floor(damage * 1.5),
      }
    }
  }

  return {
    result: ATTACK_RESULT.HIT,
    damage,
  }
}

const determineDamage = (sourceStats, isEncounter) => {
  const damageModifier = damagePoints(1, sourceStats.strength, isEncounter)
  const minDamage = sourceStats.minDamage + damageModifier
  const maxDamage = sourceStats.maxDamage + damageModifier

  const damage = Math.floor(Math.random() * (maxDamage - minDamage)) + minDamage

  return damage
}

const determineHit = (sourceStats, targetStats) =>
  Math.random() < toHitChance(sourceStats.precision, targetStats.agility)

const determineEvade = (targetStats) =>
  Math.random() < evasionChance(targetStats.agility)

const determineCritical = (sourceStats) =>
  Math.random() < criticalChance(sourceStats.precision)
