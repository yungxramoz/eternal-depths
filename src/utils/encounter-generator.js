import { ENCOUNTERS } from '../constants/encounter'
import { ENCOUNTER_BOSS } from '../constants/encounter-boss'

export const generateEncounter = (
  level,
  isBoss = false,
  random = true,
  encounter = null,
) => {
  if (random) {
    const encounters = isBoss ? ENCOUNTER_BOSS : ENCOUNTERS
    const randomIndex = Math.floor(Math.random() * encounters.length)
    const randomEncounter = encounters[randomIndex]
    encounter = randomEncounter
  }

  const name = encounter.name
  const fileName = encounter.fileName
  const idleAnimation = encounter.idleAnimation
  const stages = encounter.stages
  const style = encounter.style

  const points = level * 2
  const stats = { ...encounter.baseStats }
  generateStats(points, stats, level)

  const maxHp = stats.health * 4 + 10
  const hp = maxHp
  const minDamage = stats.strength * 2
  const maxDamageRange = isBoss ? 5 : 3
  const maxDamage = minDamage + maxDamageRange

  return {
    level,
    isBoss,
    name,
    fileName,
    idleAnimation,
    stages,
    style,
    stats,
    maxHp,
    hp,
    minDamage,
    maxDamage,
  }
}

export const encounterLevel = (stage) => {
  return Math.floor((stage - 1) / 5) + 1
}

const generateStats = (points, stats, level) => {
  const levelModification = level - 1
  stats.health += levelModification
  stats.strength += levelModification
  stats.agility += levelModification
  stats.precision += levelModification
  for (let i = 0; i < points; i++) {
    const randomStat = Math.floor(Math.random() * 4)
    switch (randomStat) {
      case 0:
        stats.health += 1
        break
      case 1:
        stats.strength += 1
        break
      case 2:
        stats.agility += 1
        break
      case 3:
        stats.precision += 1
        break
      default:
        break
    }
  }
}
