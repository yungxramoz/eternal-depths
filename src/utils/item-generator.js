import ARMOR_TYPE from '../constants/armor-type'
import RARITY from '../constants/rarity'
import WEAPON_TYPE from '../constants/weapon-type'

const items = [
  ARMOR_TYPE.HELMET,
  ARMOR_TYPE.ARMOR,
  ARMOR_TYPE.GLOVES,
  ARMOR_TYPE.GREAVES,
  WEAPON_TYPE.SWORD,
  WEAPON_TYPE.SHIELD,
]

export const generateItem = (name, item, rarity) => {
  item = item ? { ...item } : null
  if (!item) {
    item = { ...items[Math.floor(Math.random() * items.length)] }
  }
  if (!name) {
    name = item.type.charAt(0).toUpperCase() + item.type.slice(1)
  }
  if (!rarity) {
    const randomRarity = Math.random()
    if (randomRarity <= 0.65) {
      rarity = RARITY.COMMON
    } else if (randomRarity <= 0.85) {
      rarity = RARITY.RARE
    } else if (randomRarity <= 0.95) {
      rarity = RARITY.EPIC
    } else {
      rarity = RARITY.LEGENDARY
    }
  }

  const stats = { ...item.stats }
  generateStats(item.type, stats, rarity)
  item.stats = stats

  //TODO currently only swords are weapons
  item.type = item.type === 'sword' ? 'weapon' : item.type

  return { ...item, rarity, name }
}

const generateStats = (type, stats, rarity) => {
  let statIncrease
  switch (rarity) {
    case RARITY.COMMON:
      statIncrease = 1
      break
    case RARITY.RARE:
      statIncrease = 3
      break
    case RARITY.EPIC:
      statIncrease = 5
      break
    case RARITY.LEGENDARY:
      statIncrease = 8
      break
    default:
      break
  }

  if (type === 'weapon') {
    statIncrease = Math.floor(Math.random() * (statIncrease * 2)) + statIncrease
    stats.minDamage += Math.floor(Math.random() * statIncrease)
    stats.maxDamage += statIncrease - stats.minDamage
  } else {
    for (let i = 0; i < statIncrease; i++) {
      stats[randomStat(stats)] += 1
    }
  }
}

const randomStat = (stats) => {
  const statKeys = Object.keys(stats)
  return statKeys[Math.floor(Math.random() * statKeys.length)]
}
