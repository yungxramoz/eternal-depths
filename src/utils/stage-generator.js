export const generateStage = (encounter) => {
  const stages = encounter.stages
  const randomIndex = Math.floor(Math.random() * stages.length)
  return stages[randomIndex]
}
