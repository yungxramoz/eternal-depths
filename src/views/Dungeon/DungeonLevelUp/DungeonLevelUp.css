.levelup-container {
  height: 100%;
  display: flex;
  align-items: center;
  flex-direction: column;
  justify-content: center;
  text-align: center;
}

.levelup-selection-container {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 1em;
}
