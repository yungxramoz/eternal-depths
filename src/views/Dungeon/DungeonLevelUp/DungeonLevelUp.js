import React, { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import RpgButton from '../../../components/atoms/RpgButton/RpgButton'
import EditStatsPanel from '../../../components/organisms/EditStatsPanel/EditStatsPanel'
import RewardButton from '../../../components/organisms/RewardButton/RewardButton'
import RpgSeparator from '../../../components/templates/RpgSeperator/RpgSeparator'
import { RPGUI_ICON } from '../../../constants/rpgui-icon'
import {
  battleVictory,
  characterAssignAttributePoint,
  characterLearnAttack,
  characterSetAssignablePoints,
} from '../../../store/game/gameSlice'
import { generateAttack } from '../../../utils/attack'
import './DungeonLevelUp.css'
import AttackReplaceModal from './Modal/AttackReplaceModal'

const DungeonLevelUp = () => {
  const dispatch = useDispatch()
  const characterLevel = useSelector(
    (state) => state.game.character.current.level,
  )
  const characterAttacks = useSelector(
    (state) => state.game.character.current.attacks,
  )
  const [assignedPoints, setAssignedPoints] = useState({
    health: 0,
    strength: 0,
    agility: 0,
    precision: 0,
  })
  const [replaceAttackModal, setReplaceAttackModal] = useState(false)
  const availablePoints = useSelector(
    (state) => state.game.character.availableAttributePoints,
  )
  const characterStats = useSelector(
    (state) => state.game.character.current.stats,
  )
  const [attackReward] = useState(generateAttack())
  const [statRewardChoice, setStatRewardChoice] = useState(false)

  const selectAttackReward = () => {
    if (characterAttacks.length >= 3) {
      setReplaceAttackModal(true)
      return
    }
    dispatch(characterLearnAttack(attackReward))
    rewardSelected()
  }

  const selectStatReward = () => {
    dispatch(characterSetAssignablePoints(2))
    setStatRewardChoice(true)
  }

  const assignPoints = () => {
    dispatch(characterAssignAttributePoint(assignedPoints))
    rewardSelected()
  }

  const isAssignDisabled = () => {
    return (
      Object.values(assignedPoints).reduce((a, b) => a + b, 0) !==
      availablePoints
    )
  }

  const rewardSelected = () => {
    dispatch(battleVictory())
  }

  const getContent = () => {
    if (statRewardChoice) {
      return (
        <>
          <EditStatsPanel
            stats={characterStats}
            initialPoints={availablePoints}
            assignedPoints={assignedPoints}
            setAssignedPoints={setAssignedPoints}
            grey
          />
          <RpgSeparator golden={false} />
          <RpgButton
            text="Assign"
            onClick={assignPoints}
            disabled={isAssignDisabled()}
          />
        </>
      )
    }
    return (
      <>
        <p>Choose your reward:</p>
        <div className="reward-selection-container">
          <RewardButton
            icon={RPGUI_ICON.BOOK}
            title={attackReward.name}
            subtitle={`Cooldown: ${attackReward.cooldown}`}
            description={attackReward.description}
            onClick={selectAttackReward}
          ></RewardButton>
          <RpgSeparator golden />
          <RewardButton
            icon={RPGUI_ICON.POTION}
            title="Attribute"
            description="Improve 2 attributes"
            onClick={selectStatReward}
          ></RewardButton>
        </div>
      </>
    )
  }

  return (
    <>
      <div className="levelup-container">
        <h2>Level {characterLevel} reached!</h2>
        {getContent()}
      </div>
      {replaceAttackModal && (
        <AttackReplaceModal
          newAttack={attackReward}
          onClose={() => setReplaceAttackModal(false)}
          onRewardSelected={rewardSelected}
        />
      )}
    </>
  )
}

export default DungeonLevelUp
