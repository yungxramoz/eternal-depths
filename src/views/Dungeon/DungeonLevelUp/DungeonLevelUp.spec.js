import { configureStore } from '@reduxjs/toolkit'
import { fireEvent, render, screen } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import gameReducer from '../../../store/game/gameSlice'
import { generateAttack } from '../../../utils/attack'
import DungeonLevelUp from './DungeonLevelUp'

jest.mock('../../../utils/attack', () => ({
  generateAttack: jest.fn(),
}))

jest.mock('./Modal/AttackReplaceModal', () => () => (
  <div>AttackReplaceModal</div>
))

describe('DungeonLevelUp', () => {
  let store
  let attack

  beforeEach(() => {
    store = configureStore({ reducer: { game: gameReducer } })
    attack = {
      name: 'Test Attack',
      cooldown: 1,
      description: 'Test Description',
    }
    generateAttack.mockReturnValue(attack)
  })

  it('renders correctly', () => {
    render(
      <Provider store={store}>
        <DungeonLevelUp />
      </Provider>,
    )
    expect(document.body).toMatchSnapshot('DungeonLevelUp')
  })

  it('renders attribute selection when points are available', () => {
    render(
      <Provider store={store}>
        <DungeonLevelUp />
      </Provider>,
    )
    fireEvent.click(screen.getByText('Attribute'))
    expect(document.body).toMatchSnapshot('DungeonLevelUp Attribute Selection')
  })

  it('sets attribute points on attribute reward selection', () => {
    const dispatchSpy = jest.spyOn(store, 'dispatch')
    render(
      <Provider store={store}>
        <DungeonLevelUp />
      </Provider>,
    )
    fireEvent.click(screen.getByText('Attribute'))
    expect(dispatchSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        type: 'game/characterSetAssignablePoints',
        payload: 2,
      }),
    )
  })

  it('sets assigned points on attribute assign', () => {
    const dispatchSpy = jest.spyOn(store, 'dispatch')
    render(
      <Provider store={store}>
        <DungeonLevelUp />
      </Provider>,
    )
    fireEvent.click(screen.getByText('Attribute'))
    fireEvent.click(screen.getAllByRole('button')[1])
    fireEvent.click(screen.getAllByRole('button')[1])
    fireEvent.click(screen.getByText('Assign'))

    expect(screen.getByText('Available points: 0')).toBeTruthy()
    expect(dispatchSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        type: 'game/characterAssignAttributePoint',
        payload: expect.objectContaining({
          health: 2,
        }),
      }),
    )
  })

  it('learns new attack on attack reward selection', () => {
    const dispatchSpy = jest.spyOn(store, 'dispatch')
    render(
      <Provider store={store}>
        <DungeonLevelUp />
      </Provider>,
    )
    fireEvent.click(
      screen.getByRole('button', {
        name: 'Test Attack Cooldown: 1 Test Description',
      }),
    )
    expect(dispatchSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        type: 'game/characterLearnAttack',
        payload: attack,
      }),
    )
  })
  it('opens attack replace modal on attack reward selection', () => {
    store = configureStore({
      reducer: {
        game: gameReducer,
      },
      preloadedState: {
        game: {
          ...store.getState().game,
          character: {
            ...store.getState().game.character,
            current: {
              ...store.getState().game.character.current,

              attacks: [
                {
                  id: 1,
                  name: 'Attack 1',
                  currentCooldown: 0,
                },
                {
                  id: 2,
                  name: 'Attack 2',
                  currentCooldown: 0,
                },
                {
                  id: 3,
                  name: 'Attack 3',
                  currentCooldown: 0,
                },
              ],
            },
          },
        },
      },
    })
    render(
      <Provider store={store}>
        <DungeonLevelUp />
      </Provider>,
    )
    fireEvent.click(
      screen.getByRole('button', {
        name: 'Test Attack Cooldown: 1 Test Description',
      }),
    )
    expect(screen.getByText('AttackReplaceModal')).toBeTruthy()
  })
})
