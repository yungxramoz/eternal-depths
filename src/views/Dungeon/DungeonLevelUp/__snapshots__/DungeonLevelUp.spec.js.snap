// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DungeonLevelUp renders attribute selection when points are available: DungeonLevelUp Attribute Selection 1`] = `
<body>
  <div>
    <div
      class="levelup-container"
    >
      <h2>
        Level 
        1
         reached!
      </h2>
      <div
        class="rpgui-container container framed framed-golden attribute-container"
      >
        <h2>
          Attributes
        </h2>
        <p>
          Available points: 
          2
        </p>
        <hr
          class="separator"
        />
        <div
          class="attribute"
        >
          <p>
            Health
          </p>
          <div
            class="attr-controller"
          >
            <button
              class="icon-button medium "
              disabled=""
            >
              <svg
                fill="currentColor"
                id="Minus"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="2"
                  width="22"
                  x="1"
                  y="11"
                />
              </svg>
            </button>
            <p>
              1
            </p>
            <button
              class="icon-button medium "
            >
              <svg
                fill="currentColor"
                id="Plus"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="23 11 23 13 13 13 13 23 11 23 11 13 1 13 1 11 11 11 11 1 13 1 13 11 23 11"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="attribute"
        >
          <p>
            Strength
          </p>
          <div
            class="attr-controller"
          >
            <button
              class="icon-button medium "
              disabled=""
            >
              <svg
                fill="currentColor"
                id="Minus"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="2"
                  width="22"
                  x="1"
                  y="11"
                />
              </svg>
            </button>
            <p>
              1
            </p>
            <button
              class="icon-button medium "
            >
              <svg
                fill="currentColor"
                id="Plus"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="23 11 23 13 13 13 13 23 11 23 11 13 1 13 1 11 11 11 11 1 13 1 13 11 23 11"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="attribute"
        >
          <p>
            Agility
          </p>
          <div
            class="attr-controller"
          >
            <button
              class="icon-button medium "
              disabled=""
            >
              <svg
                fill="currentColor"
                id="Minus"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="2"
                  width="22"
                  x="1"
                  y="11"
                />
              </svg>
            </button>
            <p>
              1
            </p>
            <button
              class="icon-button medium "
            >
              <svg
                fill="currentColor"
                id="Plus"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="23 11 23 13 13 13 13 23 11 23 11 13 1 13 1 11 11 11 11 1 13 1 13 11 23 11"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="attribute"
        >
          <p>
            Precision
          </p>
          <div
            class="attr-controller"
          >
            <button
              class="icon-button medium "
              disabled=""
            >
              <svg
                fill="currentColor"
                id="Minus"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="2"
                  width="22"
                  x="1"
                  y="11"
                />
              </svg>
            </button>
            <p>
              1
            </p>
            <button
              class="icon-button medium "
            >
              <svg
                fill="currentColor"
                id="Plus"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="23 11 23 13 13 13 13 23 11 23 11 13 1 13 1 11 11 11 11 1 13 1 13 11 23 11"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <hr
        class="separator"
      />
      <button
        class="rpgui-button btn"
        disabled=""
      >
        Assign
      </button>
    </div>
  </div>
</body>
`;

exports[`DungeonLevelUp renders correctly: DungeonLevelUp 1`] = `
<body>
  <div>
    <div
      class="levelup-container"
    >
      <h2>
        Level 
        1
         reached!
      </h2>
      <p>
        Choose your reward:
      </p>
      <div
        class="reward-selection-container"
      >
        <button
          class="golden-button reward-button"
        >
          <div
            class="rpgui-container container framed framed-golden-2 golden-container"
          >
            <div
              class="reward-button-container"
            >
              <div
                class="rpgui-icon book reward-icon"
              />
              <div
                class="reward-button-content"
              >
                <p
                  class="reward-title"
                >
                  Test Attack
                </p>
                <p
                  class="reward-subtitle"
                >
                  <i>
                    Cooldown: 1
                  </i>
                </p>
                <p
                  class="reward-description"
                >
                  Test Description
                </p>
              </div>
            </div>
          </div>
        </button>
        <hr
          class="separator golden"
        />
        <button
          class="golden-button reward-button"
        >
          <div
            class="rpgui-container container framed framed-golden-2 golden-container"
          >
            <div
              class="reward-button-container"
            >
              <div
                class="rpgui-icon potion reward-icon"
              />
              <div
                class="reward-button-content"
              >
                <p
                  class="reward-title"
                >
                  Attribute
                </p>
                <p
                  class="reward-subtitle"
                >
                  <i />
                </p>
                <p
                  class="reward-description"
                >
                  Improve 2 attributes
                </p>
              </div>
            </div>
          </div>
        </button>
      </div>
    </div>
  </div>
</body>
`;
